src/skywallet/transaction.go

```diff
@@ -174,3 +180,147 @@ func (s *SkycoinTransactionSigner) addSignatures(msg *wire.Message) error {
 	}
 	return nil
 }
+
+// BitcoinTransactionSigner represents signing Skycoin transaction process
+// @used_in TransactionSign
+type BitcoinTransactionSigner struct {
+	Device     *Device
+	Inputs     []*messages.BitcoinTransactionInput
+	Outputs    []*messages.BitcoinTransactionOutput
+	Version    int
+	LockTime   int
+	signatures []string
+	state      int
+}
+
+// SetDevice assigns device, which will be signing
+func (s *BitcoinTransactionSigner) SetDevice(device *Device) {
+	s.Device = device
+}
+
+// Sign method signs the Bitcoin Transaction
+func (s *BitcoinTransactionSigner) Sign() ([]string, error) {
+	msg, err := s.initSigningProcess()
+
+	if err != nil {
+		return nil, err
+	}
+	index := 0
+	s.state = 0
+	for {
+		if err != nil {
+			return nil, err
+		}
+		switch msg.Kind {
+		case uint16(messages.MessageType_MessageType_TxRequest):
+			txRequest := &messages.TxRequest{}
+			err = proto.Unmarshal(msg.Data, txRequest)
+			if err != nil {
+				return nil, err
+			}
+			switch *txRequest.RequestType {
+			case messages.TxRequest_TXOUTPUT:
+				if s.state == 0 { // Sending Outputs for Confirmation
+					if len(s.Outputs)-index > 8 {
+						msg, err = s.sendOutputs(index, 8)
+						if err != nil {
+							return nil, err
+						}
+						index += 8
+					} else {
+						msg, err = s.sendOutputs(index, len(s.Outputs)-index)
+						if err != nil {
+							return nil, err
+						}
+						s.state++
+						index = 0
+					}
+				} else {
+					return nil, fmt.Errorf("protocol error: unexpected TxRequest type")
+				}
+			case messages.TxRequest_TXINPUT:
+				if s.state == 1 {
+					err = s.addSignatures(&msg)
+					if err != nil {
+						return nil, err
+					}
+					if len(s.Inputs)-index > 8 {
+						msg, err = s.sendInputs(index, 8)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						msg, err = s.sendInputs(index, len(s.Inputs)-index)
+						if err != nil {
+							return nil, err
+						}
+						s.state++
+						index = 0
+					}
+					index += 8
+				} else {
+					return nil, fmt.Errorf("protocol error: unexpected TxRequest type")
+				}
+			case messages.TxRequest_TXFINISHED:
+				if s.state == 2 {
+					err = s.addSignatures(&msg)
+					if err != nil {
+						return nil, err
+					}
+					return s.signatures, nil
+				}
+				return nil, fmt.Errorf("protocol error: unexpected TXFINISHED message")
```

Comment:
should prolly also add here what message type was received

Reply:
It is in messages.TxRequest_FINISHED case, so definetly TXFINISHED was received.